seastore: add CircularBoundedJournal

myoungwon opened the pull request:
follow up https://github.com/ceph/ceph/pull/44042

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

myoungwon commented:
@athanatos Before proceeding this work, I'd like to request a brief review to confirm I'm in the right way. The purpose of this PR is to introduce submit_transaction for multiple devices (both segment and rbm). Using existing record_t, RBM forwards extents to either circularboundedjournal or rbm.

athanatos commented:
I don't really understand the intent here.  I thought that the plan for CircularBoundedJournal was to generalize Journal into an interface with SegmentedJournal and CircularBoundedJournal implementations.  I also don't really understand why we want to submit the same transaction to multiple journals.  EPM is supposed to be responsible for submitting extents to other devices prior to commit time, but the commit itself should only take place on a single journal device.

myoungwon commented:
Ok. Then, I'll rework this PR based on the comment.

athanatos commented:
I suggest starting with the journal changes. CircularBoundedJournal should probably consume a specific offset range of an RBM interface.  An in-memory RBM implementation (like EphemeralSegmentManager) would then allow unit tests.

myoungwon commented:
I got it.

myoungwon commented:
@athanatos I made journal changes focusing on generalizing functions used in transaction_manager such as open_for_write and submit_record. Can you take a look? (need to discuss what did I miss)

athanatos commented on src/crimson/os/seastore/journal.h:
```diff
@@ -26,13 +26,101 @@ namespace crimson::os::seastore {
+  virtual ~Journal() {}
+};
+using JournalRef = std::unique_ptr<Journal>;
+
```
I don't understand the purpose of JournalManager.  What is the purpose of having more than one journal?

myoungwon replied:
If I understand segment manager and RBM design correctly, Seastore needs to handle multiple devices (segmented and rbm) ultimately. So, the purpose of this JournalManager is to forward the record to a target journal (either segmented journal or circularboundedjournal) depending on the target location. Did I miss something?

I think JournalManager can be removed for the time being.

athanatos replied:
Multiple devices, yes.  However, the purpose of a Journal is to persistently enforce ordering after a crash.  Writing different records to different journals would break those ordering guarantees (at least, without additional complications this PR doesn't seem to have).  The ExtentPlacementManager is how we're writing to devices other than the journal device, but for now there should only be a single Journal instance for any single SeaStore instance.  The existing implementation writes to a SegmentManager.  We need another implementation to allow us to journal to a RBM device instead, when one exists, so we need to generalize the existing interface to allow the users to transparently use whichever of the two implementations happens to exist.  We want to avoid other components depending specifically on SegmentJournal.

athanatos replied:
In a configuration with both a fast nvme device and a slower zns device, I'd expect that we'd have a single CBJournal on the faster nvme device with extents being split between an RBM device on the remainder of the fast nvme device and a SegmentManager on the slower ZNS device.  Does that seem reasonable?

myoungwon replied:
If I understand correctly, you mean the right way is CBJournalAllocator inherited from ExtentAllocator in the ExtentPlacementManager, not the Journal (proposed this PR)? 

If so, how we can submit the record using EPM in submit_transaction_direct in case of CBJournal. something like epm->submit_record(), not journal->submit_record()?

> In a configuration with both a fast nvme device and a slower zns device, I'd expect that we'd have a single CBJournal on the faster nvme device with extents being split between an RBM device on the remainder of the fast nvme device and a SegmentManager on the slower ZNS device. Does that seem reasonable?

Yes.

athanatos replied:
I'm not sure I understand what you mean.  I'd have expected that we could simply change Journal to an interface with the existing implementation renamed to SegmentedJournal inheriting from Journal.  We'd then add another implementation, CBJournal, also inheriting from Journal.  On startup, either a CBJournal or a SegmentedJournal would be created and passed to all components as Journal.

myoungwon replied:
I think that is what I intended for this PR except for JournalManager.  If JournalManager is removed in this PR,  does these commits already address your comments (even though the commits needs to be improved)? ExtentPlacementManager can take over a role of JournalManager later.

athanatos replied:
Replacing JournalManager dependencies with with Journal should work. TransactionManager should depend on Journal rather than JournalManager as you've said.  EPM should depend on the general Journal interface rather than SegmentJournal.  

We won't need EPM to replace the JournalManager.  Writing to devices other than the journal device doesn't mean writing to another journal, it means doing an out-of-line (ool) allocation on another device prior to the journal commit.  That capability is partly already in place, see ExtentPlacementManager::delayed_alloc_or_ool_write and related code and @cyx1231st's pending PR https://github.com/ceph/ceph/pull/44809.  Writing to an RBM device rather than than a SegmentManager device would mean extending the existing ExtentAllocator interface to allocate from an RBM device as the SegmentedAllocator does from a SegmentedDevice.

My current thinking is that we probably won't ever need to permit more than one journal device.  If we ever choose to permit multiple journal devices, we'll do so under the Journal interface.

I think the next step after this refactor will be to modify TransactionManager/EPM to be able to force nearly all extents to be written ool (I think that's the goal for the CBJournal?) with the existing SegmentedAllocator and a CBJournal with SegmentCleaner updates to ensure the CBJournal has enough space.  Existing unit tests where appropriate should be updated to also test that configuration.  The step after that would be to integrate an RBM allocator to do ool writes to RBM devices.

Does that make sense?

cyx1231st replied:
> I think the next step after this refactor will be to modify TransactionManager/EPM to be able to force nearly all extents to be written ool (I think that's the goal for the CBJournal?) with the existing SegmentedAllocator and a CBJournal with SegmentCleaner updates to ensure the CBJournal has enough space.

IIUC this is referring to the TODOs about:
1. https://github.com/ceph/ceph/blob/32da0e0ecb70224d005deb13bda8808f2665a811/src/crimson/os/seastore/extent_placement_manager.h#L313
2. https://github.com/ceph/ceph/blob/32da0e0ecb70224d005deb13bda8808f2665a811/src/crimson/os/seastore/extent_placement_manager.h#L318

Forcing all extents to be written ool means RBM will be able to decouple all extents out of journal records (i.e. inlining), so that extents can be decoupled out of journaling & trimming processes with their allocations decided by the RBM allocator. And journals can be bounded to a reserved space effectively if without inline extents. Hopefully with this direction we could reuse the existing Journal logic as much as possible.

This will also benefit segmented manager as we will be able to classify cold and hot extents more effectively towards better segmented GC.

BTW I'm working on 2 with further refactors to generalize EPM.

Does this also make sense to you @athanatos ?

myoungwon replied:
> We won't need EPM to replace the JournalManager. Writing to devices other than the journal device doesn't mean writing to another journal, it means doing an out-of-line (ool) allocation on another device prior to the journal commit. That capability is partly already in place, see ExtentPlacementManager::delayed_alloc_or_ool_write and related code and @cyx1231st's pending PR https://github.com/ceph/ceph/pull/44809. Writing to an RBM device rather than than a SegmentManager device would mean extending the existing ExtentAllocator interface to allocate from an RBM device as the SegmentedAllocator does from a SegmentedDevice.

I got it.

> My current thinking is that we probably won't ever need to permit more than one journal device. If we ever choose to permit multiple journal devices, we'll do so under the Journal interface.

I agree.

> I think the next step after this refactor will be to modify TransactionManager/EPM to be able to force nearly all extents to be written ool (I think that's the goal for the CBJournal?) with the existing SegmentedAllocator and a CBJournal with SegmentCleaner updates to ensure the CBJournal has enough space. Existing unit tests where appropriate should be updated to also test that configuration. The step after that would be to integrate an RBM allocator to do ool writes to RBM devices.

Seems reasonable to me.

athanatos replied:
@cyx1231st Yeah, I think that's mostly right.

athanatos commented on src/crimson/os/seastore/transaction_manager.cc:
```diff
@@ -88,7 +88,7 @@ TransactionManager::mount_ertr::future<> TransactionManager::mount()
   return segment_cleaner->init_segments().safe_then(
     [this](auto&& segments) {
-    return journal->replay(
+    return journal->as_seg_journal().replay(
```
All journals need to be able to replay, why cast to SegmentedJournal first?

myoungwon replied:
I added a commit to generalize replay().

athanatos commented on src/crimson/os/seastore/transaction_manager.h:
```diff
@@ -551,7 +551,7 @@ class TransactionManager : public SegmentCleaner::ExtentCallbackInterface {
 	*sm,
 	*lba_manager,
-	*journal,
+	journal->as_seg_journal(),
```
It's not obvious to me why this needs to be a SegmentJournal.  SegmentedAllocator is using this instance to get the sequence id, not to perform the write.  Passing a CBJournal would be fine, and necessary in cases where are writing ool extents to a zns device in a configuration with a CBJournal.

myoungwon replied:
Rebased. Due to recent changes, the journal is no longer the input argument.

myoungwon commented:
I added a commit to address the comment with regard to replay(). Plus, rebase is done.

athanatos commented:
I reworked the refactor portion a bit: https://github.com/ceph/ceph/pull/45089.  If that looks ok to you, we can merge that and then rebase this one.

athanatos commented:
I will get to this later this week.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,286 @@
+ *
+ */
+
+class CBJournal : public Journal {
```
CircularBoundedJournal

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,286 @@
+#include <list>
+
+
+namespace crimson::os::seastore {
```
crimson::os::seastore::journal to match source code path and segmented_journal.h

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+    ).safe_then([](auto super) {
+      return mkfs_ertr::now();
+    }).handle_error(
+      crimson::ct_error::enoent::handle([this, &config, start_addr] (auto) 
```
](auto)

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+    assert(config.block_size == device->get_block_size());
+    return read_super(start_addr
+    ).safe_then([](auto super) {
+      return mkfs_ertr::now();
```
It's not clear to me what's going on here.  Looks like if read_super() suceeds, we return success?  Presumably this is to allow mkfs() to be idempotent?  Should we be checking the returned super against the config uuid?  Certainly, the semantics of this method need to be explained in a comment on the method in the header.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+	  encode(head, bl);
+	} catch (ceph::buffer::error &e) {
+	  logger().debug("unable to encode super block from underlying deivce");
+	  return mkfs_ertr::now();
```
Is this pathway actually possible?  This can't happen unless the encode implementation is actually broken, I don't think returning success is the right answer if so.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+  try {
+    bl = encode_super();
+  } catch (ceph::buffer::error &e) {
+    logger().debug("unable to encode super block from underlying deivce");
```
This isn't a device problem, this can only happen if there's an actual software bug in the implementation of encode.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+}
+
+CBJournal::write_ertr::future<>
+CBJournal::sync_super()
```
write_super()

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+    }).safe_then([this](auto p) mutable {
+      auto &[head, bl] = *p;
+      header = head;
+      logger().debug(" super : {} ", header);
```
Throughout this PR, let's use the LOG_PREFIX, DEBUG etc macros in seastore/logging.h.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+
+  auto bptr = bufferptr(ceph::buffer::create_page_aligned(to_write.length()));
+  auto iter = to_write.cbegin();
+  iter.copy(to_write.length(), bptr.c_str());
```
See bufferlist::rebuild_aligned

athanatos replied:
Also, this should be using device_write_bl, I think.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+CBJournal::write_ertr::future<> CBJournal::append_record(
+  ceph::bufferlist bl,
+  rbm_abs_addr addr)
+{
```
This method sequentially writes out the two portions of the record.  Is there any reason not to submit the two writes concurrently?

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+    set_used_size(get_used_size() + (header.end - get_written_to()));
+    set_written_to(get_start_addr());
+    if (encoded_size > get_available_size()) {
+      return crimson::ct_error::erange::make();
```
This should be impossible since it gets checked earlier in the method?

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+  auto write_length = length < get_block_size() ? get_block_size() : length;
+  auto bptr = bufferptr(ceph::buffer::create_page_aligned(write_length));
+  auto iter = bl.cbegin();
+  iter.copy(bl.length(), bptr.c_str());
```
See bufferlist::rebuild_aligned

myoungwon replied:
added rebuild_aligned above.

athanatos replied:
Yes, which means that these lines are not necessary.  bufferlist::rebuild_aligned updates bl to contain only aligned bufferptrs -- you can write them directly without allocating/copying a new one.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+	cursor_addr,
+	header.device_id);
+      return read_record(cursor_paddr
+      ).safe_then([this, &cursor_addr, &d_handler] (auto ret) {
```
](

Please fix any other examples of this formatting.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,286 @@
+    return get_total_size() - get_used_size();
+  }
+
+  void update_applied_to(rbm_abs_addr addr, uint32_t len) {
```
This doesn't appear to have any callers.

myoungwon replied:
This is used by unittest-seastore-cbjournal for a test purpose because there is no cleaner implementation for now. I explained applied_to as below.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,286 @@
+  rbm_abs_addr get_applied_to() const {
+    return header.applied_to;
+  }
+  void set_applied_to(rbm_abs_addr addr) {
```
This only gets called from update_applied_to, which does not seem to have any callers.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,286 @@
+    rbm_abs_addr start_offset;      // start offset of CBJournal
+    rbm_abs_addr last_committed_record_base;
+    rbm_abs_addr written_to;
+    rbm_abs_addr applied_to;
```
I'm not sure what applied_to is meant to denote.

myoungwon replied:
This represents the cursor which indicates the last record---moved to RBM most recently---position in CircularBoundedJournal. The cleaner will take the role---move records from CircularBoundedJournal to RBM, but the cleaner for CBjournal is not implemented yet. I added this to confirm that the journal work as circular journal.

athanatos replied:
Is the intention to use applied_to during replay to reduce the number of entries you need to apply?  How is this different from the tail of the journal?

myoungwon replied:
Yes, it is the same as the tail of the journal.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+}
+
+CBJournal::read_record_ret CBJournal::read_record(paddr_t off)
+{
```
This method has some formatting problems.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,573 @@
+
+Journal::replay_ret CBJournal::replay(
+    delta_handler_t &&delta_handler)
+{
```
I'm confused by applied_to, so I'll look more closely at replay() on my next pass.

myoungwon commented:
I added explanations to describe why applied_to is needed as above, and added commits to address your comments---each commit would match a comment.

myoungwon commented:
@athanatos ping.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,569 @@
+  : device(device), path(path) {}
+
+CircularBoundedJournal::mkfs_ret
+CircularBoundedJournal::mkfs(mkfs_config_t& config)
```
const mkfs_config_t &config or mkfs_config_t config -- don't pass by mutable reference without a good reason.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,569 @@
+    head.cur_segment_seq = 0;
+    head.start = start_addr;
+    head.end = end_addr;
+    head.device_id = config.device_id;
```
Most of this should be part of the default constructor for cbj_header_t.  Seems like we can only update the things that are actually non-default.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,295 @@
+   *
+   */
+
+  struct cbj_header_t {
```
This needs a default constructor that initializes every field.  For every field without a default constructor, add an initializer:
```
  uint64_t magic = CBJOURNAL_MAGIC
  uuid_d uuid;
  ...
```

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,569 @@
+
+CircularBoundedJournal::close_ertr::future<> CircularBoundedJournal::close()
+{
+  return write_super(
```
Under what conditions does the super block actually change?  If it actually does change, is it ok to defer updating it until close?

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,569 @@
+    DEBUG(
+      "initialize header block in CircularBoundedJournal, length {}",
+      bl.length());
+    return device_write_bl(start_addr, bl
```
"super" refers to the encoded header of type cbj_header_t, right?  Let's not use "super" to refer to this, let's do write_super->write_header and read_super->read_header.  We use "super" elsewhere already and this code is inconsistent between those two names.

Also, use the write_header helper here, someone grepping the code base for "write_header" should find all of the sites where we perform header writeout.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,569 @@
+   * read records from last applied record prior to written_to, and replay
+   */
+  LOG_PREFIX(CircularBoundedJournal::replay);
+  if (get_last_committed_record_base() == get_applied_to()) {
```
I'm probably misunderstanding the semantics here, but this seems wrong.  replay() gets invoked prior to open_for_write() and I don't see anywhere else where the header gets read.

Even if the header does get read prior to this point, I think there's still a problem.  I see set_last_committed_record_base() being called at the end of submit_record prior to completion -- that seems right.  However, set_last_committed_record_base() updates the in-memory copy of header without performing a write-out.  If close() isn't called (perhaps because the node in question was power-cycled), rereading the header will result in a badly out of date header with a value of last_committed_record_base which is badly out of date.

For that reason, I'm somewhat confused by the purpose of cbj_header_t::last_committed_record_base.  Since you can't rely on a clean shutdown, its value must necessarily be out of date unless you plan on rewriting the header for every record.  Is it meant as a way to detect broken/corrupted devices?

athanatos replied:
I would expect that replay() would find written_to and last_committed_record_base from the last readable record rather than relying on the value in the header.

myoungwon replied:
> I'm probably misunderstanding the semantics here, but this seems wrong. replay() gets invoked prior to open_for_write() and I don't see anywhere else where the header gets read.

Added.


> Even if the header does get read prior to this point, I think there's still a problem. I see set_last_committed_record_base() being called at the end of submit_record prior to completion -- that seems right. However, set_last_committed_record_base() updates the in-memory copy of header without performing a write-out. If close() isn't called (perhaps because the node in question was power-cycled), rereading the header will result in a badly out of date header with a value of last_committed_record_base which is badly out of date.

Right. To address the issue you mentioned, I added a commit (https://github.com/ceph/ceph/pull/44233/commits/77e3799a3c94c7bd890a111021230ce4bd9c0853). The intention here is that CBJournal makes applied_to (the tail of journal) persistent whenever applying---apply the content of records in CBjournal to RBM---occurs, not every writes (submit_record). Since replay() will be started from appliied_to, it can recover last_committed_records_base and written_to correctly if we maintain applied_to safely. Note that replay() try to read record sequentially until it finds invalid records. Make sense?
In addition, I have plan to implement applied_to update---storing applied_to persistent---with cleaner changes (future work item).

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,295 @@
+
+    rbm_abs_addr start_offset;      // start offset of CircularBoundedJournal
+    rbm_abs_addr last_committed_record_base;
+    rbm_abs_addr written_to;
```
See above, written_to and last_committed_record_base must necessarily be out of date unless they get rewritten on every write.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,295 @@
+    uint64_t magic;
+    uuid_d uuid;
+    uint64_t block_size; // aligned with block_size
+    uint64_t size;   // max length of journal
```
This can be recovered as cbj_header_t::end - cbj_header_t::begin, right?

myoungwon replied:
Right. addressed.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,295 @@
+    uuid_d uuid;
+    uint64_t block_size; // aligned with block_size
+    uint64_t size;   // max length of journal
+    uint64_t used_size;  // current used_size of journal
```
This has the same problem as written_to and last_committed_record_base.  Additionally, it can be recovered from start_offset and written_to (once you find the last valid record during replay).

myoungwon replied:
please see the last commit.

athanatos commented:
Ok, I've done another pass.  See my above comments, I think I'm misunderstanding the commit semantics between replay and submit_record.

myoungwon commented on src/crimson/os/seastore/cache.cc:
```diff
@@ -1173,7 +1173,13 @@ void Cache::complete_commit(
+	addr_types_t::RANDOM_BLOCK) {
+	i->set_paddr(final_block_start.add_offset(
+	  i->get_paddr().as_seg_paddr().get_segment_off()));
+      }
```
I think we need to discuss this. I suppose that there are two ways to use blk_paddr with existing seastore implementation.

1) allocate blk_paddr directly after epm.alloc_new_extent
2) update blk_paddr after submit_record 

IMO, 1) probably causes racing problem (performance degradation) and does not guarantee sequential allocation within a transaction context because alloc_new_extent can be called by multiple threads. So, without real allocation at alloc_new_extent(), it seems that address allocation after transaction is made is better way. Am I wrong?

athanatos replied:
That allocation pathway is for ool extents.  It seems to me that extents within the CBJournal should be addressed pre-commit as with a SegmentedJournal -- as record-relative.  Post commit, CBJournal will return the actual blk_paddr addresses and we'll update the extent addresses as we do for any journal commit.

myoungwon replied:
Ok. The commit (seastore: enable cbjournal when running unittest_transaction_manager) works as your comment.

myoungwon commented:
I posted commits to address your comment---integrate CBJournal test into unittest-transaction-manager--including comments here. I think we need to discuss CBjournal with unittest-transaction-manager.

myoungwon commented:
@athanatos I posted commits to address your comment---integrate CBJournal test into unittest-transaction-manager--including comments here. I think we need to discuss CBjournal with unittest-transaction-manager (please see https://github.com/ceph/ceph/pull/44233/commits/7193491a79e822c44e89a299566a88e50bdf0efc).

athanatos commented:
Please add a short comment at the top of the CBJournal class explaining the two important sides of the commit protocol:
1) what happens at commit time
2) what happens at replay time

Ensure that you mention any relevant metadata fields or dependent commits.  Please also ensure that all header fields have a single line explanation.

myoungwon commented:
@athanatos I added more explanation (please see the last commit). Please take a look.

athanatos commented:
I'll get to this in the next week or so.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -30,7 +30,27 @@ using NVMeBlockDevice = nvme_device::NVMeBlockDevice;
+ *
+ * - Commit time
+ * After submit_record is done, last_committed_record_base and written_to are
+ * advanced---last_committed_record_base indicates the last committed record and
```
"are advanced" are these fields in the header?  Do you rewrite the header on each commit?

myoungwon replied:
Sorry for confusing you. CBJournal will rewrite the fields, but these fields are in memory (without storing the fields permanently).

After submit_record is done, last_committed_record_base and written_to are increased (these are in memory values)

athanatos replied:
Why?  What purpose do they serve?

myoungwon replied:
Removed last_committed_record_base and rework is done based on your comments. See the updated commits.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -30,7 +30,27 @@ using NVMeBlockDevice = nvme_device::NVMeBlockDevice;
+ *
+ * - Replay time
+ * At replay time, CBJournal begins to replay records in CBjournal by reading
+ * records from applied_to. Then, CBJournal examines whether the records is valid
```
Is applied_to a header field?  When is the ondisk field updated?

myoungwon replied:
> Is applied_to a header field? When is the ondisk field updated?

Yes. 
After the apply work---applying the records in CBJournal to RBM---is done.

athanatos replied:
I don't think this is actually correct.  Applying records to RBM means that *those* deltas won't need to be replayed, but the CBJournal will in general include deltas for records on SegmentedDevices as well.  I don't think applied_to is actually relevant to the replay semantics of CBJournal, it simply restricts which RBM write-ahead deltas can be ignored.  Even then, I think its only an optimization.

myoungwon replied:
I didn't consider that CBjournal can include records on SegmentedDevices. I added journal_tail when replaying records. See the last updated commit.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -30,7 +30,27 @@ using NVMeBlockDevice = nvme_device::NVMeBlockDevice;
+ * At replay time, CBJournal begins to replay records in CBjournal by reading
+ * records from applied_to. Then, CBJournal examines whether the records is valid
+ * one by one, at which point last_committed_record_base and written_to are
+ * recovered if the valid record is founded. Note that only applied_to is managed
```
What makes a record "valid"?

myoungwon replied:
there are two conditions: 1) decoding without an exception, 2) correct sequence number (this should be less than prior valid record).

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+ *
+ * - Commit time
+ * After submit_record is done, last_committed_record_base and written_to are
+ * increased (these are in-memory values)---last_committed_record_base indicates
```
What purpose does last_committed_record_base serve?

myoungwon replied:
Removed.

athanatos commented on src/crimson/os/seastore/btree/fixed_kv_node.h:
```diff
@@ -218,7 +218,12 @@ struct FixedKVInternalNode
+	if (base.get_addr_type() == addr_types_t::SEGMENT) {
+	  updated = base.add_relative(i->get_val());
+	} else {
+	  updated = base.add_offset(i->get_val().as_seg_paddr().get_segment_off());
```
add_relative has well-defined behavior for both segmented and rbm addresses.  Fix paddr_t::add_relative to behave correctly regardless of the type so that we don't need to add these branches whenever we need to use add_relative.

athanatos commented on src/crimson/os/seastore/cache.cc:
```diff
@@ -989,7 +989,8 @@ record_t Cache::prepare_record(
       auto sseq = NULL_SEG_SEQ;
       auto stype = segment_type_t::NULL_SEG;
-      if (cleaner != nullptr) {
+      if (cleaner != nullptr && i->get_paddr().get_addr_type() ==
```
This isn't handled correctly for rbm addresses yet.  Just do assert(i->get_paddr().get_addr_type() == addr_types_t::SEGMENT)

myoungwon replied:
Since `i` is a member of mutual_block_list, i->get_paddr() here can be the rbm address.
Consequently, adding assert will trigger assert during unittest-transaction-manager.

athanatos commented on src/crimson/os/seastore/cache.cc:
```diff
@@ -1175,7 +1176,13 @@ void Cache::complete_commit(
+	addr_types_t::RANDOM_BLOCK) {
+	i->set_paddr(final_block_start.add_offset(
+	  i->get_paddr().as_seg_paddr().get_segment_off()));
+      }
```
Same comment, the goal paddr_t representing both segmented and rbm addresses was specifically to avoid needing to do this kind of branch.  Update add_relative to behave correctly for rbm addresses as well.

athanatos commented on src/crimson/os/seastore/lba_manager/btree/lba_btree_node.cc:
```diff
@@ -39,7 +39,11 @@ void LBALeafNode::resolve_relative_addrs(paddr_t base)
+	val.paddr = base.add_relative(val.paddr);
+      } else {
+	val.paddr = base.add_offset(val.paddr.as_seg_paddr().get_segment_off());
+      }
```
See above.

athanatos commented on src/crimson/os/seastore/seastore_types.cc:
```diff
@@ -572,8 +572,14 @@ try_decode_deltas(
+      } else if (record_block_base.get_addr_type() ==
+	  addr_types_t::RANDOM_BLOCK) {
+	auto& blk_addr = record_block_base.as_blk_paddr();
+	blk_addr.set_block_off(blk_addr.get_block_off() + i.len);
```
Once more, we are adding an offset to an address. Add a paddr_t method to do this correctly for both types transparently.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -642,10 +642,10 @@ struct seg_paddr_t : public paddr_t {
     assert(!base.is_block_relative());
-    seg_paddr_t& s = base.as_seg_paddr();
-    if (is_block_relative())
+    if (is_block_relative()) {
```
This is actually wrong.  Addresses within an extent which refer to other extents within the same record will be block relative whether the extents are in a CBJournal or a SegmentedJournal.  Fix/add a paddr_t::add_block_relative method to behave correctly whether base is random or segmented.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -1122,7 +1122,12 @@ class __attribute__((packed)) phy_tree_root_t {
+	root_addr = base.add_record_relative(_root_addr);
+      } else {
+	// RANDOM_BLOCK
+	root_addr = base.add_offset(_root_addr.as_seg_paddr().get_segment_off());
```
See above.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -1710,6 +1715,9 @@ inline paddr_t paddr_t::add_record_relative(paddr_t o) const {
   PADDR_OPERATION(addr_types_t::SEGMENT, seg_paddr_t, maybe_relative_to(o))
+  if (get_addr_type() == addr_types_t::RANDOM_BLOCK) {
+    return *this;
+  }
```
Is there a reason you can't use the PADDR_OPERTION macro?

athanatos commented on src/crimson/os/seastore/segment_cleaner.cc:
```diff
@@ -480,6 +480,9 @@ segment_id_t SegmentCleaner::allocate_segment(
 void SegmentCleaner::update_journal_tail_target(journal_seq_t target)
 {
+  if (target.offset.get_addr_type() == addr_types_t::RANDOM_BLOCK) {
+    return;
```
This isn't correct.  Make this an assert until you add support for it.  (Unless that creates a problem for testing?)

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -633,12 +633,21 @@ class SegmentCleaner : public SegmentProvider {
+      journal_tail_committed = JOURNAL_SEQ_NULL;
+      journal_head = JOURNAL_SEQ_NULL;
+      return;
+    }
```
SegmentCleaner or whatever it eventually becomes will eventually need to perform journal trimming even if the journal is CBJournal.  This seems wrong.

athanatos commented on src/crimson/os/seastore/transaction_manager.cc:
```diff
@@ -538,14 +538,20 @@ TransactionManager::get_extent_if_live_ret TransactionManager::get_extent_if_liv
 TransactionManager::~TransactionManager() {}
 
-TransactionManagerRef make_transaction_manager(bool detailed)
+TransactionManagerRef make_transaction_manager(bool detailed, bool cbjournal)
```
Instead of passing this as a param, can we read it out of the config?

athanatos commented:
This PR seems to have a bunch of modifications that just disable portions of the SegmentCleaner that don't work yet, mainly the journal trimming portions.  Those portions apply just as much to CircularBoundedJournal as SegmentedJournal, right?

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -633,12 +633,21 @@ class SegmentCleaner : public SegmentProvider {
   void set_journal_head(journal_seq_t head) {
+    if (head.offset.get_addr_type() == addr_types_t::RANDOM_BLOCK) {
+      return;
+    }
```
See above, not sure why this needs to be disabled.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -690,6 +701,8 @@ class SegmentCleaner : public SegmentProvider {
     if (!init_complete)
       return;
+    if (addr.get_addr_type() != addr_types_t::SEGMENT)
+      return;
```
This and mark_space_used make sense since SegmentCleaner won't be tracking space usage for RBM devices.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+    uint64_t flag = 0;       // represent features (reserved)
+    uint8_t csum_type = 0;   // type of checksum algoritghm used in cbj_header_t
+    uint64_t csum = 0;       // checksum of entire cbj_header_t
+    uint32_t cur_segment_seq = 0;
```
?

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+
+    uint64_t flag = 0;       // represent features (reserved)
+    uint8_t csum_type = 0;   // type of checksum algoritghm used in cbj_header_t
+    uint64_t csum = 0;       // checksum of entire cbj_header_t
```
rename to header_checksum

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+    rbm_abs_addr applied_to = 0;
+
+    uint64_t flag = 0;       // represent features (reserved)
+    uint8_t csum_type = 0;   // type of checksum algoritghm used in cbj_header_t
```
This should really be an enum.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+    uint32_t error = 0;      // reserved
+
+    // start offset of CircularBoundedJournal in the device (start + header length)
+    rbm_abs_addr start_offset = 0;
```
Rename to journal_tail

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+    // start offset of CircularBoundedJournal in the device (start + header length)
+    rbm_abs_addr start_offset = 0;
+    // start address where last committed record is written
+    rbm_abs_addr last_committed_record_base = 0;
```
We don't update this on every write, so it will almost always be less the actual last committed_record_base.  If the intention is to use it as a conservative bound, change to be journal_head_lower_bound.  Otherwise, just remove it.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+    uuid_d uuid;
+    uint64_t block_size = 0; // aligned with block_size
+    uint64_t size = 0;   // max length of journal
+    uint64_t used_size = 0;  // current used_size of journal
```
This doesn't make sense to me.  It duplicates journal_tail/last_committed_record_base and is, in general, wrong for the same reason that last_committed_record_base is.  Remove it.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+  struct cbj_header_t {
+    uint64_t magic = CBJOURNAL_MAGIC;
+    uuid_d uuid;
+    uint64_t block_size = 0; // aligned with block_size
```
Is there any reason this would differ from RBM::block_size?  Change the comment to "block size of underlying device"

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+    // start address where last committed record is written
+    rbm_abs_addr last_committed_record_base = 0;
+    // start address where the newest record will be written
+    rbm_abs_addr written_to = 0;
```
This is duplicative of last_committed_record_base and will in general be out of date for the same reason.  Doesn't seem useful.

athanatos replied:
As I mentioned before, this field will, in general, be out of date since you do not write the header out on commit.  For this reason, leaving it in the header is dangerous as someone might write code that assumes that the header encoded on disk has a valid written_to value.  replay() immediately overwrites this value after calling open_for_write() -- there's no obvious reason to write it into the header at all.  Make this a member of CircularBoundedJournal directly.

athanatos replied:
This hasn't been addressed?

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+    uint32_t cur_segment_seq = 0;
+
+    rbm_abs_addr start = 0; // start address of CircularBoundedJournal
+    rbm_abs_addr end = 0;   // start address of CircularBoundedJournal
```
How are these different from start_offset and size?

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,341 @@
+    uint64_t block_size = 0; // aligned with block_size
+    uint64_t size = 0;   // max length of journal
+    uint64_t used_size = 0;  // current used_size of journal
+    uint32_t error = 0;      // reserved
```
There isn't actually a reason to reserve fields.  The denc encode/decode system has versioning that enables adding fields later.  Let's just stick to the fields that are actually used.

athanatos commented:
I'll get back to this next week.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -297,10 +297,7 @@ CircularBoundedJournal::write_ertr::future<> CircularBoundedJournal::device_writ
-  auto bptr = bufferptr(ceph::buffer::create_page_aligned(write_length));
-  auto iter = bl.cbegin();
-  iter.copy(bl.length(), bptr.c_str());
+  auto bptr = bl.begin().get_current_ptr();
```
Look at the implementation of rebuild_aligned.  There may be multiple bufferptrs, but each is guarranteed to be aligned.  Is there a device method that can submit multiple write for each aligned buffer?

athanatos replied:
If you really need it to be a single buffer, add a bufferptr bufferlist::get_aligned_bufferptr() const method.

myoungwon replied:
I added writev commit to address your comment.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,316 @@
+    // address to represent where last appllied record is written
+    rbm_abs_addr applied_to = 0;
+
+    uint64_t flag = 0;       // represent features (reserved)
```
Still not useful to reserve bits in the header.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,316 @@
+  NVMeBlockDevice* device;
+  std::string path;
+  WritePipeline *write_pipeline = nullptr;
+  bool init = false;
```
Rename to initialized.  Add comment:

```
/** 
 * initialized
 * 
 * true after open_read_header, set to false in close().  Indicates that device is open
 * and in-memory header is valid.
 */

athanatos replied:
Hasn't been renamed?

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,568 @@
+}
+
+CircularBoundedJournal::open_for_write_ret
+CircularBoundedJournal::open_for_write(rbm_abs_addr start)
```
This will not be used interchangeably with open_for_write(), rename it to open_device_read_header.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,568 @@
+CircularBoundedJournal::open_for_write(rbm_abs_addr start)
+{
+  LOG_PREFIX(CircularBoundedJournal::open_for_write);
+  if (init) {
```
ceph_assert(!init);

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,568 @@
+}
+
+CircularBoundedJournal::open_for_write_ret CircularBoundedJournal::open_for_write()
+{
```
See above, this *must* be called after replay, and therefore after open_device_read_header.

```
    ceph_assert(initialized);
    paddr_t paddr = convert_abs_addr_to_paddr(
      get_written_to(),
      header.device_id);
    return open_for_write_ret(
      open_for_write_ertr::ready_future_marker{},
      journal_seq_t{
	cur_segment_seq,
	paddr
      });
```

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,568 @@
+  auto r_size = record_group_size_t(record.size, get_block_size());
+  auto encoded_size = r_size.get_encoded_length();
+  if (get_written_to() +
+      ceph::encoded_sizeof_bounded<record_group_header_t>() > get_journal_end()) {
```
This check/behavior belongs in append_record, it's a special case of generally needing to split a record across the end and beginning of the journal.

athanatos commented on src/crimson/os/seastore/btree/fixed_kv_node.h:
```diff
@@ -218,7 +218,7 @@ struct FixedKVInternalNode
     for (auto i: *this) {
       if (i->get_val().is_relative()) {
-	auto updated = base.add_relative(i->get_val());
+	paddr_t updated = base.add_relative(i->get_val());
```
Is this necessary?

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -677,6 +676,33 @@ struct blk_paddr_t : public paddr_t {
+    if (o.get_addr_type() == addr_types_t::SEGMENT) {
+      // segment addr is allocated when alloc_new_extent is called.
+      // But, if random block device is used,
+      // this eventually can be converted to blk addr after submit_record().
```
This is misleading, all record-relative or block-relative addrs are represented internally as segmented.  I don't think it's possible to have a non-segmented relative addr.  I think you can ceph_assert(o.get_addr_type() == addr_types_t::SEGMENT).

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -677,6 +676,33 @@ struct blk_paddr_t : public paddr_t {
+    return add_offset(off);
+  }
+
+  paddr_t maybe_relative_to(paddr_t base) const {
```
Worth adding a comemnt

```
  // all blk_paddr_t are absolute, relative addrs are always segmented
```

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -641,7 +642,9 @@ class SegmentCleaner : public SegmentProvider {
     assert(journal_head == JOURNAL_SEQ_NULL || head >= journal_head);
     journal_head = head;
-    segments.update_written_to(head.offset);
+    if (head.offset.get_addr_type() == addr_types_t::SEGMENT) {
```
This should also be an assert.  Once you add support, it will be entirely fine to call set_journal_head on addresses for the CBJournal.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -747,6 +754,10 @@ class SegmentCleaner : public SegmentProvider {
     return space_tracker->equals(tracker);
   }
 
+  void set_disable_trim(bool val){
```
Space between ) and {

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -589,6 +589,7 @@ class SegmentCleaner : public SegmentProvider {
 
   SegmentSeqAllocatorRef ool_segment_seq_allocator;
 
+  bool disable_trim = false; // for test
```
Let's add more detail to the comment

```
/**
 * disable_trim
 * 
 * added to enable unit testing of CircularBoundedJournal before proper support is added to
 * SegmentCleaner.  Should be removed once proper support is added. TODO
 */

athanatos commented on src/crimson/os/seastore/transaction_manager.cc:
```diff
@@ -538,14 +538,21 @@ TransactionManager::get_extent_if_live_ret TransactionManager::get_extent_if_liv
+  } else {
+    journal = journal::make_circularbounded(
+      nullptr, "");
+    segment_cleaner->set_disable_trim(true);
```
Add a log line here:

```
ERROR("disabling journal trimming since support for CircularBoundedJournal hasn't been added yet");
```

athanatos commented on src/test/crimson/seastore/test_transaction_manager.cc:
```diff
@@ -71,8 +72,22 @@ struct transaction_manager_test_t :
   }
 
+  bool for_segmented() {
+    std::string j_type = GetParam();
```
return j_type == "segmented";

athanatos commented on src/test/crimson/seastore/test_transaction_manager.cc:
```diff
@@ -581,7 +601,8 @@ struct transaction_manager_test_t :
 		  return tm->alloc_extent<TestBlock>(
-		    *(t.t), L_ADDR_MIN, size
+		    *(t.t), L_ADDR_MIN, size,
+		    for_segmented() ? placement_hint_t::HOT : placement_hint_t::REWRITE
```
This doesn't belong at this interface level.  Remove the p_hint argument from alloc_hint and add a default_placement_hint config to the tm_make_config_t.

athanatos commented on src/test/crimson/seastore/test_transaction_manager.cc:
```diff
@@ -1026,12 +1061,30 @@ TEST_F(tm_single_device_test_t, find_hole_assert_trigger)
+    "segmented",
+    "circularbounded"
+  )
+);
```
Nice!  I think this test infrastructure is a really good start.

athanatos commented on src/test/crimson/seastore/transaction_manager_test_state.h:
```diff
@@ -109,12 +121,15 @@ class EphemeralTestState {
         sec_sm.reset();
       }
+      if (j_type == journal_type::CIRCULARBOUNDED_JOURNAL) {
+	rb_device.reset();
```
Just do this unconditionally, it'll be harmless.

athanatos commented on src/test/crimson/seastore/transaction_manager_test_state.h:
```diff
@@ -58,11 +62,19 @@ class EphemeralTestState {
 
-  seastar::future<> tm_setup() {
+  seastar::future<> tm_setup(journal_type j_type = journal_type::SEGMENT_JOURNAL) {
+    this->j_type = j_type;
```
j_type can be local, I think.  No need for it to be part of EphemeralTestState

athanatos commented on src/test/crimson/seastore/transaction_manager_test_state.h:
```diff
@@ -137,8 +152,15 @@ class TMTestState : public EphemeralTestState {
   virtual void _init() override {
-    tm = make_transaction_manager(true);
+    tm_make_config_t config;
+    config.cbjournal = j_type == journal_type::SEGMENT_JOURNAL ? false : true;
```
```
config.cbjournal = j_type == journal_type::CIRCULARBOUNDED_JOURNAL;
```

athanatos commented on src/crimson/tools/store_nbd/tm_driver.cc:
```diff
@@ -131,7 +131,8 @@ seastar::future<bufferlist> TMDriver::read(
 void TMDriver::init()
 {
-  tm = make_transaction_manager(false /* detailed */);
+  tm_make_config_t config {false, false};
```
No space between config and {

athanatos commented on src/crimson/os/seastore/journal.h:
```diff
@@ -18,6 +18,11 @@ class NVMeBlockDevice;
 class SegmentManagerGroup;
 class SegmentProvider;
 
+enum class journal_type {
```
journal_type_t

Type name are always either CamelCase (generally for complex types) or snake_case_t (simple types like enums).

athanatos commented:
Ok, I think we're getting close.  The transaction_manager unit tests are a great start to testing CircularBoundedJournal integration with the rest of SeaStore.

myoungwon commented:
@athanatos I addressed your comments.

athanatos commented:
Needs to be rebased

myoungwon commented:
@athanatos Done.

athanatos commented on src/test/crimson/seastore/test_transaction_manager.cc:
```diff
@@ -774,7 +798,8 @@ TEST_F(tm_single_device_test_t, allocate_lba_conflict)
       SIZE,
-      'a');
+      'a',
+      tm_config.default_placement_hint);
```
I mean that this argument is passed into alloc_extent identically everywhere in this file.  Remove the hint argument from alloc_extent entirely and set the default there.

athanatos commented on src/crimson/os/seastore/transaction_manager.h:
```diff
@@ -294,13 +294,14 @@ class TransactionManager : public SegmentCleaner::ExtentCallbackInterface {
     laddr_t laddr_hint,
-    extent_len_t len) {
+    extent_len_t len,
+    placement_hint_t hint = placement_hint_t::HOT) {
```
Instead of passing a hint argument here, just set placement_hint in the else branch below based on the configured value you added.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,320 @@
+    // address to represent where last appllied record is written
+    rbm_abs_addr applied_to = 0;
+
+    checksum_t header_checksum = 0;       // checksum of entire cbj_header_t
```
This doesn't seem to be used.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,320 @@
+    // start offset of CircularBoundedJournal in the device
+    rbm_abs_addr journal_tail = 0;
+    // start address where the newest record will be written
+    rbm_abs_addr written_to = 0;
```
(sorry, I left this comment yesterday, but it doesn't seem to show up on the main page)

As I mentioned before, this field will, in general, be out of date since you do not write the header out on commit. For this reason, leaving it in the header is dangerous as someone might write code that assumes that the header encoded on disk has a valid written_to value. replay() immediately overwrites this value after calling open_for_write() -- there's no obvious reason to write it into the header at all. Make this a member of CircularBoundedJournal directly.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.h:
```diff
@@ -0,0 +1,320 @@
+   * true after open_device_read_header, set to false in close().
+   * Indicates that device is open and in-memory header is valid.
+   */
+  bool init = false;
```
This still needs to be renamed to initialized.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,570 @@
+
+CircularBoundedJournal::open_for_write_ret CircularBoundedJournal::open_for_write()
+{
+  if (init) {
```
Rename init to initialized, add ceph_assert(initialized), remove the open_device_read_header(CBJOURNAL_START_ADDRESS); branch

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -0,0 +1,570 @@
+  auto r_size = record_group_size_t(record.size, get_block_size());
+  auto encoded_size = r_size.get_encoded_length();
+  if (get_written_to() +
+      ceph::encoded_sizeof_bounded<record_group_header_t>() > get_journal_end()) {
```
(this got missed from yesterday)



This check/behavior belongs in append_record, it's a special case of generally needing to split a record across the end and beginning of the journal.

myoungwon replied:
The purpose of this line is to set written_to to the start address of CBjournal because of there is not enough space (< records_group_header_t) between written_to---address to write a record---and the end of CBjournal. So, the record will be append to the beginning of CBjournal. 
Note that this line does not intend to split a record the end and beginning of the CBjournal.

athanatos replied:
Ok.

athanatos commented:
Some comments from yesterday got missed, and I clarified a few.

myoungwon commented:
I addressed your comments and add a comment to explain the if condition. Please take a look. I'll rebase this after the recent unittest-transaction-manager fail is resolved.

athanatos commented:
Needs to be rebased again.

athanatos commented on src/crimson/os/seastore/journal/circular_bounded_journal.cc:
```diff
@@ -326,6 +332,14 @@ CircularBoundedJournal::read_header(rbm_abs_addr start)
+    ceph_le32 recorded_crc_le;
+    decode(recorded_crc_le, bliter);
+    uint32_t recorded_crc = recorded_crc_le;
+    ceph_assert(test_crc == recorded_crc);
```
Add useful logging here.

athanatos commented:
@cyx1231st I'm inclined to merge this on monday.  The disable_trim behavior is a bit clumsy, but it enables some initial unit testing of this machinery that we can expand with better SegmentCleaner support for CBJournal.  Let me know if you'd like some additional time to review it.

myoungwon commented:
@athanatos Rebased and added the logging.

myoungwon commented:
@athanatos Rebased.

athanatos commented:
@myoungwon I'm sorry, I didn't get to this in time today and something else is conflicting with it.  Rebase one more time, ping me, and I'll get it merged before we see another conflict.  Sorry again!

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -787,16 +811,31 @@ struct journal_seq_t {
+	return -1;
+      }
+    } else {
+      // offset.get_addr_type() != other.get_addr_type()
```
We really want journal_seq_t values to work correctly between types, I'll submit a PR later to fix this. https://tracker.ceph.com/issues/55705

athanatos commented on src/crimson/os/seastore/transaction_manager.h:
```diff
@@ -34,6 +34,20 @@
+    placement_hint_t::HOT;
+  static tm_make_config_t get_default() {
+    return tm_make_config_t {
+      true,
```
This should only be set to true for tests.  I'll submit a PR shortly to fix this https://tracker.ceph.com/issues/55706
